token.go

```diff
@@ -0,0 +1,95 @@
+package sdk
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// expiryDelta determines how much earlier a token should be considered
+// expired than its actual expiration time. It is used to avoid late
+// expirations due to client-server time mismatches.
+const expiryDelta = 10 * time.Second
+
+// Token represents an OpenFaaS ID token
+type Token struct {
+	// IDToken is the OIDC access token that authorizes and authenticates
+	// the requests.
+	IDToken string
+
+	// Expiry is the expiration time of the access token.
+	//
+	// A zero value means the token never expires.
+	Expiry time.Time
+}
+
+// Expired reports whether the token is expired.
+func (t *Token) Expired() bool {
+	if t.Expiry.IsZero() {
+		return false
+	}
+
+	return t.Expiry.Round(0).Add(-expiryDelta).Before(time.Now())
+}
+
+type tokenJSON struct {
+	AccessToken string `json:"access_token"`
+	IDToken     string `json:"id_token"`
+
+	TokenType string `json:"token_type"`
+
+	ExpiresIn int `json:"expires_in"`
+}
+
+func (t *tokenJSON) expiry() (exp time.Time) {
+	if v := t.ExpiresIn; v != 0 {
+		return time.Now().Add(time.Duration(v) * time.Second)
+	}
+	return
+}
+
+// Exchange an ID Token for OpenFaaS token
+func ExchangeIDToken(tokenURL, rawIDToken string) (*Token, error) {
+	v := url.Values{}
+	v.Set("grant_type", "urn:ietf:params:oauth:grant-type:token-exchange")
+	v.Set("subject_token_type", "urn:ietf:params:oauth:token-type:id_token")
+	v.Set("subject_token", rawIDToken)
+
+	u, _ := url.Parse(tokenURL)
+
+	req, err := http.NewRequest(http.MethodPost, u.String(), strings.NewReader(v.Encode()))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := io.ReadAll(res.Body)
+	res.Body.Close()
```

Comment:
You need to move this under the if check otherwise it will panic at runtime if there is an error and body is nil.

We also use defer res.Body.Close() everywhere else, perhaps we should be consistent?

Reply:
Updated this.